File: backend/Backend/src/main/java/com/sgw/backend/entity/Voucher.java
package com.sgw.backend.entity;

import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

import com.fasterxml.jackson.annotation.JsonIgnore;

@EqualsAndHashCode(callSuper = true)
@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@DiscriminatorValue("voucher")
public class Voucher extends Billable {

    private String voucherName;

    private BigDecimal voucherValue;

    private String voucherCode;

    private LocalDate voucherExpiryDate;

    private VoucherStatusEnum voucherStatusEnum;

    private String voucherApplicableVenue;

    // Mcdonalds voucher implementation variables
    // Activate then usable (15min)
    private boolean isActive;

    private LocalDateTime activationTime;

    @JsonIgnore
    @ManyToOne(optional = false)
    @JoinColumn(name = "voucher_student_id")
    private Student voucherStudent;

    @JsonIgnore // adding this just in case... i foresee many problems
    @ManyToOne
    private Wallet voucherWallet;

    @Override
    public String getLabel() {
        return voucherName;
    }
}

File: student_frontend/lib/model/voucher.dart
class Voucher {
  final int billableId;
  final double billablePrice;
  final String voucherName;
  final double voucherValue;
  final String voucherCode;
  final String voucherExpiryDate;
  final String voucherStatusEnum;
  final String? activationTime;
  final String voucherApplicableVenue;
  final bool active;

  Voucher({
    required this.billableId,
    required this.billablePrice,
    required this.voucherName,
    required this.voucherValue,
    required this.voucherCode,
    required this.voucherExpiryDate,
    required this.voucherStatusEnum,
    this.activationTime,
    required this.voucherApplicableVenue,
    required this.active,
  });

  factory Voucher.fromJson(Map<String, dynamic> json) {
    return Voucher(
      billableId: json['billableId'],
      billablePrice: json['billablePrice'].toDouble(),
      voucherName: json['voucherName'],
      voucherValue: json['voucherValue'].toDouble(),
      voucherCode: json['voucherCode'],
      voucherExpiryDate: json['voucherExpiryDate'],
      voucherStatusEnum: json['voucherStatusEnum'],
      activationTime: json['activationTime'],
      voucherApplicableVenue: json['voucherApplicableVenue'],
      active: json['active'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'billableId': billableId,
      'billablePrice': billablePrice,
      'voucherName': voucherName,
      'voucherValue': voucherValue,
      'voucherCode': voucherCode,
      'voucherExpiryDate': voucherExpiryDate,
      'voucherStatusEnum': voucherStatusEnum,
      'activationTime': activationTime,
      'voucherApplicableVenue': voucherApplicableVenue,
      'active': active,
    };
  }
}
